atom-jobs/atom-ut-codecov.groovy

```diff
@@ -0,0 +1,213 @@
+                            serverPath: '/mnt/ci.pingcap.net-nfs/git', readOnly: false),
+            ],
+    ) {
+        node(label) {
+            println "debug command:\nkubectl -n ${namespace} exec -ti ${NODE_NAME} bash"
+            body()
+        }
+    }
+}
+
+
+lines_coverage_rate = ""
+
+run_with_pod {
+    container("golang") {
+        try {
+            def ws = pwd()
+
+            stage("${TASK_NAME}") {
+                println "${TASK_NAME}"
+            }
+
+            stage("Download binary from fileserver") {
+                sh '''
+                    wget -q -c ${CACHE_CODE_FILESERVER_URL} -O - | tar -xz
+                '''
+            }
+
+            credentialList =[]
+            varStrings = SECRET_VARS.split(",")
+            for (varString in varStrings) {
+                keyAndValue = varString.split(":")
+                if (keyAndValue.length == 2) {
+                    credentialList.push(string(credentialsId: keyAndValue[0], variable: keyAndValue[1]))
+                }
+            }
+
+            stage("Test") {
+                dir("${ws}/${REPO}") {
+                    sh TEST_CMD
+                }
+            }
+
+            stage("Codecov") {
+                // upload coverage report to codecov
+                // read coverage rate from codecov api
+                dir("${ws}/${REPO}") {
+                    // coverage file: test_report/*coverage.out
+                    withCredentials(credentialList) {
+                        timeout(5) {
+                            if (PULL_REQUEST_ID != null && PULL_REQUEST_ID != "") {
+                                sh """
+                                curl -LO ${FILE_SERVER_URL}/download/cicd/ci-tools/codecov
+                                chmod +x codecov
+                                ./codecov -f 'coverage_report/*.coverage'  -t ${CODECOV_TOKEN} -C ${COMMIT_ID} -P ${PULL_REQUEST_ID} -b ${BUILD_NUMBER}
+                                """
+                            } else {
+                                sh """
+                                curl -LO ${FILE_SERVER_URL}/download/cicd/ci-tools/codecov
+                                chmod +x codecov
+                                ./codecov -f 'coverage_report/*.coverage' -t ${CODECOV_TOKEN} -C ${COMMIT_ID} -b ${BUILD_NUMBER} -B ${BRANCH}
+                                """
+                            }
+                        }
+                        sleep(time:120,unit:"SECONDS")
+                        def response = httpRequest Authorization: CODECOV_API_TOKEN, url: "https://codecov.io/api/gh/pingcap/tidb/commit/${COMMIT_ID}"
+                        println('Status: '+response.status)
+                        def obj = readJSON text:response.content
+                        if (response.status == 200) {
+                            println(obj.commit.totals)
+                            lines_coverage_rate = "Lines coverage ${obj.commit.totals.c.toFloat().round(2)}%."
```

Comment:
Should we fail build when not reach coverage rate？

Reply:
if ut test failed, current coverage is not correct, so we not display coverage when ut failed